itest/lnd_forward_interceptor_test.go

```diff
@@ -344,6 +347,226 @@ func testForwardInterceptorBasic(ht *lntest.HarnessTest) {
+		ctx := context.Background()
+		dbgInfo, err := carol.RPC.LN.GetDebugInfo(
+			ctx, &lnrpc.GetDebugInfoRequest{},
+		)
+		require.NoError(ht, err, "failed to get Carol node debug info")
+
+		for _, logEntry := range dbgInfo.Log {
+			if logEntryCheck(logEntry) {
+				return true
+			}
+		}
+
+		return false
+	}, defaultTimeout, time.Second)
+
+	// Cancel the context, which will disconnect Bob's interceptor.
+	cancelBobInterceptor()
+
+	// Make sure all goroutines are finished.
+	select {
+	case <-done:
+	case <-time.After(defaultTimeout):
+		require.Fail(ht, "timeout waiting for sending payment")
+	}
+
+	// Assert that the payment was successful.
+	var preimage lntypes.Preimage
+	copy(preimage[:], invoice.RPreimage)
+	ht.AssertPaymentStatus(alice, preimage, lnrpc.Payment_SUCCEEDED)
+
+	// Finally, close channels.
+	ht.CloseChannel(alice, cpAB)
+	ht.CloseChannel(bob, cpBC)
+}
+
+// testForwardInterceptorFirstHopRecords
+func testForwardInterceptorFirstHopRecords(ht *lntest.HarnessTest) {
+	// Initialize the test context with 3 connected nodes.
+	ts := newInterceptorTestScenario(ht)
+
+	alice, bob, carol := ts.alice, ts.bob, ts.carol
+
+	// Open and wait for channels.
+	const chanAmt = btcutil.Amount(300000)
+	p := lntest.OpenChannelParams{Amt: chanAmt}
+	reqs := []*lntest.OpenChannelRequest{
+		{Local: alice, Remote: bob, Param: p},
+		{Local: bob, Remote: carol, Param: p},
+	}
+	resp := ht.OpenMultiChannelsAsync(reqs)
+	cpAB, cpBC := resp[0], resp[1]
+
+	// Make sure Alice is aware of channel Bob=>Carol.
+	ht.AssertTopologyChannelOpen(alice, cpBC)
+
+	// Connect an interceptor to Bob's node.
+	// bobInterceptor, cancelBobInterceptor := bob.RPC.HtlcInterceptor()
+
+	req := &lnrpc.Invoice{ValueMsat: 1000}
+	addResponse := carol.RPC.AddInvoice(req)
+	invoice := carol.RPC.LookupInvoice(addResponse.RHash)
+
+	sendReq := &routerrpc.SendPaymentRequest{
+		PaymentRequest: invoice.PaymentRequest,
+		TimeoutSeconds: int32(wait.PaymentTimeout.Seconds()),
+		FeeLimitMsat:   noFeeLimitMsat,
+		FirstHopCustomRecords: map[uint64][]byte{
+			65537: []byte("test"),
+		},
+	}
+	stream := alice.RPC.SendPayment(sendReq)
+	ht.AssertPaymentStatusFromStream(stream, lnrpc.Payment_SUCCEEDED)
+
+	// Check that the modified fields were reported in Bob's log.
+	// targetLogPrefixStr := "Received UpdateAddHTLC("
+
+	// Formulate custom records target log string.
+	var (
+		cr  record.CustomSet = make(record.CustomSet)
+		buf bytes.Buffer
+	)
+
+	cr[65537] = []byte("test")
+
+	err := cr.Encode(&buf)
+	require.NoError(ht, err, "failed to encode custom records")
+	customRecordsBytes := buf.Bytes()
+
+	targetCustomRecordsStr := fmt.Sprintf(
+		"custom_records_blob=%x", customRecordsBytes,
+	)
+
+	logEntryCheck := func(logEntry string) bool {
+		return strings.Contains(logEntry, "custom_records_blob") &&
+			strings.Contains(logEntry, targetCustomRecordsStr)
+	}
+
+	require.Eventually(ht, func() bool {
+		ctx := context.Background()
+		dbgInfo, err := bob.RPC.LN.GetDebugInfo(
```

Comment:
Since there's not yet support to retrieve the custom blob from the interceptor, we'll assert that bob receives the first hop records by peeking at his logs